server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1484,6 +1561,11 @@ public void addShardFailureCallback(Consumer<ShardFailure> onShardFailure) {
     public void checkIdle(long inactiveTimeNS) {
         Engine engineOrNull = getEngineOrNull();
         if (engineOrNull != null && System.nanoTime() - engineOrNull.getLastWriteNanos() >= inactiveTimeNS) {
+            if (active.get() == false) {
+                // We refresh when transitioning to an inactive state to make
+                // it easier to cache the store size.
+                refresh("transition to inactive");
```

Comment:
-1 we should not refresh anything except of the internal reader. Visibility guarantees are important.